flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/failover/AdaptedRestartPipelinedRegionStrategyNG.java

```diff
@@ -107,11 +107,16 @@ protected void restartTasks(final Set<ExecutionVertexID> verticesToRestart) {
 
 		FutureUtils.assertNoException(
 			cancelTasks(verticesToRestart)
-				.thenRunAsync(resetAndRescheduleTasks(globalModVersion, vertexVersions), executionGraph.getJobMasterMainThreadExecutor())
+				.thenApply(ignored -> createResetAndRescheduleTasksRunnable(globalModVersion, vertexVersions))
+				.thenAccept(restartAction -> executionGraph.getRestartStrategy()
```

Comment:
I don't think the `.thenApply` stage is needed. Also, the `.thenAccept` stage breaks the level of abstraction. I'd extract the body into a new function. Consider this structure:

```
	FutureUtils.assertNoException(
			cancelTasks(verticesToRestart)
				.thenRun(scheduleRestart(globalModVersion, vertexVersions)) // strictly speaking should be thenRunAsync?
				.handle(failGlobalOnError()));
	}

	private Runnable scheduleRestart(final long globalModVersion, final Set<ExecutionVertexVersion> vertexVersions) {
		final RestartStrategy restartStrategy = executionGraph.getRestartStrategy();
		return () -> restartStrategy.restart(
			resetAndRescheduleTasks(globalModVersion, vertexVersions),
			executionGraph.getJobMasterMainThreadExecutor()
		);
	}

	private RestartCallback resetAndRescheduleTasks(final long globalModVersion, final Set<ExecutionVertexVersion> vertexVersions) {
		...
```